### PR TITLE
feat(decoder): add image header cache

### PR DESCRIPTION
**Pull request by W-Mai**
### Description of the feature or fix

Add image decoder header cache

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html)
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by W-Mai** [comment]
this pr relies on other prs
merge order:

1. #5402 
2. #5405
3. #5404
4. #5407

**Review by XuNeo — APPROVED** [review]
Looks good to me.

File: src/draw/lv_image_decoder.c [hunk 1]
```diff
@@ -62,6 +75,14 @@ void _lv_image_decoder_init(void)
         .create_cb = NULL,
         .free_cb = (lv_cache_free_cb_t)image_decoder_cache_free_cb,
     });
+
+    // FIXME: max_count should be configurable
```
Comment (terry0012) [inline src/draw/lv_image_decoder.c]: why not a config?
If a new config is added, image header cache code should move out of `#if LV_CACHE_DEF_SIZE > 0` ?

File: src/draw/lv_image_decoder.c [hunk 2]
```diff
@@ -72,106 +93,56 @@ void _lv_image_decoder_deinit(void)
...
-    }
-
-    if(dsc->src_type == LV_IMAGE_SRC_FILE)
-        lv_free((void *)dsc->src);
+    /*
+     * We assume that if a decoder can get the info, it can open the image.
+     * If decoder open failed, free the source and return error.
+     * If decoder open succeed, add the image to cache if enabled.
+     * */
+    lv_result_t res = dsc->decoder->open_cb(dsc->decoder, dsc);
```
Comment (terry0012) [inline src/draw/lv_image_decoder.c:152]: where to free `dsc->src` if src is LV_IMAGE_SRC_FILE  and `open_cb` return failed ?
should each `open_cb`  of image decoder do it?  I don't see any place do that.

File: src/draw/lv_image_decoder.c (hunk 1 above)
Comment (W-Mai) [reply src/draw/lv_image_decoder.c]: It is necessary to add the independent control macro of `header cache`, but there is no time to add it. I'll add it later. 😁

File: src/draw/lv_image_decoder.c (hunk 2 above)
Comment (W-Mai) [reply src/draw/lv_image_decoder.c:152]: Regardless of success or failure, decoder `open` and `close` appear in pairs, with src allocated in `open` and released in `close`.

File: src/draw/lv_image_decoder.c (hunk 1 above)
Comment (W-Mai) [reply src/draw/lv_image_decoder.c]: Done!

File: src/draw/lv_image_decoder.c (hunk 2 above)
Comment (XuNeo) [reply src/draw/lv_image_decoder.c:152]: Will you move the unified `cache_free_cb` to lv_image_decoder.c?

Comment (W-Mai) [reply src/draw/lv_image_decoder.c:152]: Haven't yet.

Create a new pr to deal this thing separately. 😁

I haven't forgotten it. It's on my to-do list.

File: src/draw/lv_image_decoder.c [hunk 3]
```diff
@@ -321,6 +299,70 @@ lv_draw_buf_t * lv_image_decoder_post_process(lv_image_decoder_dsc_t * dsc, lv_d
...
+
+#if LV_IMAGE_HEADER_CACHE_DEF_CNT > 0
+    if(decoder) {
+        if(src_type == LV_IMAGE_SRC_FILE) search_key.src = lv_strdup(src);
+        search_key.decoder = decoder;
+        search_key.header = *header;
+        entry = lv_cache_add(img_header_cache_p, &search_key, NULL);
+
+        if(entry == NULL) {
+            if(src_type == LV_IMAGE_SRC_FILE) lv_free((void *)src);
```
Comment (PGNetHun) [inline src/draw/lv_image_decoder.c]: I think the duplicated string should be freed up (otherwise there is a memory leak):
`lv_free(search_key.src)`

**Review by PGNetHun — CHANGES_REQUESTED** [review]

File: src/draw/lv_image_decoder.c (hunk 3 above)
Comment (W-Mai) [reply src/draw/lv_image_decoder.c]: You're right, my mistakes. 😭

Comment (W-Mai) [reply src/draw/lv_image_decoder.c]: Done

**Review by PGNetHun — APPROVED** [review]

**Review by kisvegabor — APPROVED** [review]
Thank you for this update and the reviews!

**Comment by W-Mai** [comment]
Wait, these patches have dependencies. Joining this patch alone will make it difficult for the code to roll back review and so on. Can we merge them one by one?



> this pr relies on other prs merge order:
> 
> 1. [feat(image_decoder): remove `args` for `open_cb` #5402](https://github.com/lvgl/lvgl/pull/5402)
> 2. [fix(decoder_dsc): replace `decoder_dsc.header` with `decoder_dsc.decoded->header` #5405](https://github.com/lvgl/lvgl/pull/5405)
> 3. [feat(image_decoder): simplify `lv_image_decoder_get_info/open` workflow #5404](https://github.com/lvgl/lvgl/pull/5404)
> 4. [feat(decoder): add image header cache #5407](https://github.com/lvgl/lvgl/pull/5407)

**Comment by kisvegabor** [comment]
Ok, I'll merge them in this order, but please resolve merge conflicts.

**Comment by kisvegabor** [comment]
I reverted #5407

**Comment by W-Mai** [comment]
In fact, their interdependence is a superposition relationship. If I fix the conflict, it will become an empty patch, so my suggestion is that we quickly revert this patch, and then merge it in order, and then resubmit the image header cache. If it's troublesome, I can ask @FASTSHIFT for help.😁

**Comment by W-Mai** [comment]
> I reverted #5407

Oh, You are one step faster!

**Comment by kisvegabor** [comment]
This one still need to be updated: https://github.com/lvgl/lvgl/pull/5407

**Comment by W-Mai** [comment]
> This one still need to be updated: #5404

Done, waiting for the CI.

**Comment by kisvegabor** [comment]
Please open this PR again to merge it again :slightly_smiling_face:

**Comment by W-Mai** [comment]
> Please open this PR again to merge it again 🙂

It started CI. 
1. #5420

**Pull request merged**